src/routes/(blog-article)/profiling-the-torrust-bittorrent-udp-tracker/+page.md

```diff
@@ -0,0 +1,328 @@
+update_one_torrent_in_parallel: Avg/AdjAvg: (3.542378ms, 3.435695ms)
+add_multiple_torrents_in_parallel: Avg/AdjAvg: (15.651172ms, 15.651172ms)
+update_multiple_torrents_in_parallel: Avg/AdjAvg: (4.368189ms, 4.257572ms)
+
+tokio::sync::RwLock<std::collections::BTreeMap<InfoHash, Arc<tokio::sync::Mutex<Entry>>>>
+add_one_torrent: Avg/AdjAvg: (111ns, 109ns)
+update_one_torrent_in_parallel: Avg/AdjAvg: (6.590677ms, 6.808535ms)
+add_multiple_torrents_in_parallel: Avg/AdjAvg: (16.572217ms, 16.30488ms)
+update_multiple_torrents_in_parallel: Avg/AdjAvg: (4.073221ms, 4.000122ms)
+```
+
+</CodeBlock>
+
+## Profiling With Valgrind
+
+"Valgrind is an instrumentation framework for building dynamic analysis tools". In fact it's a suite of tools. The tool we are going to use is `callgrind` which collects data that can be later visualized with `kcachegrind`.
+
+Valgrind, coupled with Kcachegrind, offers a powerful profiling solution for in-depth analysis of BitTorrent tracker performance. By simulating cache behavior and collecting call graphs, developers gain valuable insights into code execution dynamics and potential optimizations.
+
+In order to profile the UDP tracker you need to:
+
+1. Build and run the tracker for profiling.
+2. Make requests to the tracker while it's running.
+
+Build and run the binary for profiling with:
+
+<CodeBlock lang="terminal">
+
+```terminal
+RUSTFLAGS='-g' cargo build --release --bin profiling \
+   && export TORRUST_TRACKER_PATH_CONFIG="./share/default/config/tracker.udp.benchmarking.toml" \
+   && valgrind \
+     --tool=callgrind \
+     --callgrind-out-file=callgrind.out \
+     --collect-jumps=yes \
+     --simulate-cache=yes \
+     ./target/release/profiling 60
+```
+
+</CodeBlock>
+
+<Callout type="info">
+
+**NOTICE:** You should make requests to the services you want to profile. For example, using the [Aquatic UDP load test](./benchmarking-the-torrust-bittorrent-tracker).
+
+</Callout>
+
+After running the tracker with `valgrind` it generates a file called `callgrind.out`
+that you can open with `kcachegrind`.
+
+<CodeBlock lang="terminal">
+
+```terminal
+kcachegrind callgrind.out
+```
+
+</CodeBlock>
+
+<Image src="/images/posts/profiling-the-torrust-bittorrent-udp-tracker/kcachegrind-screenshot.png" alt="Visualizing profiling data with kcachegrind" />
+
+You have information like:
+
+- How many times a function is called (`Called` column).
+- The total amount of time spent running a function (`Incl.` column).
+- The total amount of time spent running a function excluding child functions (`Self` column).
+- Etcetera.
+
+But one of the most interesting things it's the interactive "Call Graph" where you can navigate through functions seeing how much time they spend.
+
+The main problem with this tool is the asynchronous code. Functions executed in tokio spawned tasks are not shown directly. Since the Torrust Tracker make an intense use of async code that makes it harder to reach conclusions from the tool.
```

Comment:
orthography:
..... . Since the Torrust Tracker "makes" intense  use of async code.....